File: app/src/main/java/com/friendlypos/Recibos/util/TotalizeHelperRecibos.java
package com.friendlypos.Recibos.util;

import android.util.Log;

import com.friendlypos.Recibos.activity.RecibosActivity;
import com.friendlypos.Recibos.modelo.recibos;
import com.friendlypos.login.util.SessionPrefes;

import java.util.List;

import io.realm.Realm;

import static io.realm.internal.SyncObjectServerFacade.getApplicationContext;

public class TotalizeHelperRecibos {

    private static final double IVA = 13.0;
    String customer;
    RecibosActivity activity;
    private static double productosDelBonus = 0;
    SessionPrefes session;


    public TotalizeHelperRecibos(RecibosActivity activity) {
        this.activity = activity;
        session = new SessionPrefes(getApplicationContext());
    }

    public void destroy() {
        activity = null;
    }


    public void totalizeRecibos(List<recibos> pivotList) {
        for (recibos p : pivotList) {
            totalize(p);
        }
    }

    private void totalize(final recibos currentPivot) {

        Double totalPago = currentPivot.getTotal();

        Double total = 0.0;
        total = totalPago;


        Double totalPagado = currentPivot.getPaid();

        Double totalPagar = 0.0;
        totalPagar = totalPago - totalPagado;

        Log.d("TOTALTODOS", total +"");

        activity.setTotalizarTotal(total);
        activity.setTotalizarCancelado(totalPagar);

        final String facturaId = currentPivot.getInvoice_id();

        final Realm realm2 = Realm.getDefaultInstance();
        final Double finalTotalPagar = totalPagar;

        realm2.executeTransaction(new Realm.Transaction() {

            @Override
            public void execute(Realm realm2) {
                recibos recibo_actualizado = realm2.where(recibos.class).equalTo("invoice_id", facturaId).findFirst();


                recibo_actualizado.setPorPagar(finalTotalPagar);

                realm2.insertOrUpdate(recibo_actualizado);
                realm2.close();

                Log.d("ACTRECIBOPAGAR", recibo_actualizado + "");

            }
        });



    }
}

File: app/src/main/java/com/friendlypos/Recibos/util/TotalizeHelperRecibos.kt
package com.friendlysystemgroup.friendlypos.Recibos.util

import android.content.Context
import android.util.Log
import com.friendlysystemgroup.friendlypos.Recibos.activity.RecibosActivity
import com.friendlysystemgroup.friendlypos.Recibos.modelo.recibos
import com.friendlysystemgroup.friendlypos.login.util.SessionPrefes
import io.realm.Realm

/**
 * Clase auxiliar para totalizar recibos
 */
class TotalizeHelperRecibos(
    private var activity: RecibosActivity?,
    context: Context
) {
    private var customer: String? = null
    private var facturaId: String? = null
    private val session: SessionPrefes = SessionPrefes(context)

    /**
     * Libera recursos cuando ya no se necesita la instancia
     */
    fun destroy() {
        activity = null
    }

    /**
     * Totaliza una lista de recibos
     * @param pivotList Lista de recibos a totalizar
     */
    fun totalizeRecibos(pivotList: List<recibos>) {
        if (pivotList.isEmpty()) {
            Log.d("TotalizeHelper", "No hay recibos para totalizar")
            return
        }
        
        // Totalizar cada recibo individualmente
        pivotList.forEach { recibo ->
            totalize(recibo)
        }

        // Actualizar el monto por pagar en la base de datos
        actualizarMontoPorPagar()
    }
    
    /**
     * Actualiza el monto por pagar en la base de datos
     */
    private fun actualizarMontoPorPagar() {
        if (facturaId.isNullOrEmpty()) {
            Log.e("TotalizeHelper", "ID de factura no disponible para actualización")
            return
        }
        
        val finalTotalPagar = activity?.getTotalizarCancelado() ?: 0.0
        
        val realm = Realm.getDefaultInstance()
        try {
            realm.executeTransaction { r ->
                val reciboActualizado = r.where(recibos::class.java)
                    .equalTo("invoice_id", facturaId)
                    .findFirst()
                    
                reciboActualizado?.let {
                    it.porPagar = finalTotalPagar
                    r.insertOrUpdate(it)
                    Log.d("ReciboActualizado", "Recibo con ID $facturaId actualizado: ${it.porPagar}")
                } ?: Log.e("TotalizeHelper", "No se encontró el recibo con ID $facturaId")
            }
        } finally {
            realm.close()
        }
    }

    /**
     * Totaliza un recibo individual
     * @param currentPivot Recibo a totalizar
     */
    private fun totalize(currentPivot: recibos) {
        val totalPago = currentPivot.total
        val totalPagado = currentPivot.paid
        val totalPagar = totalPago - totalPagado

        Log.d("Totalizando", "Total: $totalPago, Pagado: $totalPagado, Por pagar: $totalPagar")

        activity?.apply {
            setTotalizarTotal(totalPago)
            setTotalizarCancelado(totalPagar)
        }

        facturaId = currentPivot.invoice_id
    }

    companion object {
        private const val IVA = 13.0
    }
}
